File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.java
class Solution {
   
     public List<String> letterCombinationsRec(String p,String up){
       
		if(up.isEmpty()) {
			ArrayList<String> list = new ArrayList<String>();
			list.add(p);
			return list;
		}
		ArrayList<String> list = new ArrayList<String>();
		int d = up.charAt(0) - '0'; 
		int i = (d-2)*3;
		int len = i+3;
         if(d>7) {
			i=(d-2)*3 + 1;
			len++;
		}
		if(d==7 || d==9) {
			len = i+4;
		}
		
		for(;i<len;i++)
		{
//			ArrayList<String> list = new ArrayList<String>();
			char ch = (char)('a'+i);
			list.addAll(letterCombinationsRec(p+ch,up.substring(1)));
            // System.out.println(list);
			
		}
		
		return list;
	}
     
    
    public List<String> letterCombinations(String digits) {
        if(digits.length()==0)
            return new ArrayList<String>();
        return letterCombinationsRec("",digits);
    }
}

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution:
     def letterCombinations(self, digits: str) -> List[str]:
        map_ = {
            '2' : 'abc',
            '3' : 'def',
            '4' : 'ghi',
            '5' : 'jkl',
            '6' : 'mno',
            '7' : 'pqrs',
            '8' : 'tuv',
            '9' : 'wxyz'
        }
        result = []
        def make_combinations(i, cur):
            if i == len(digits):
                if len(cur) > 0:
                    result.append(''.join(cur))
                return
            for ch in map_[digits[i]]:
                cur.append(ch)
                make_combinations(i+1, cur)
                cur.pop()
        make_combinations(0, [])
        return result    